syntax="proto3";
//指定当前proto文件的语法是3系列版本


// Usage: protoc [OPTION] PROTO_FILES
// --cpp_out=OUT_DIR           Generate C++ header and source.

// protoc --cpp_out=. msg.proto
// protoc        protobuf的可执行文件   
// --cpp_out=.   生成文件存放在当前目录下  
// msg.proto     源文件名字


//指定生成代码后，相关结构定义所在的命名空间
package pb;

//无关选项，用于客户端
option csharp_namespace="Pb";


// 1 登录
// 5 退出
message SyncPid 
{
    //玩家ID
    int32 Pid=1;
    //玩家姓名
    string Username=2;
}

message Player
{
    //玩家ID
    int32 Pid=1;
    //坐标
    Position P=2;
    //玩家姓名
    string Username=3;
}

// 6 周围玩家
message SyncPlayers
{
    //数组
    //嵌套多个子消息类型Player的消息：多个玩家
    repeated Player ps=1;
}

//3 新位置
message Position
{
    float X=1;
    float Y=2;
    float Z=3;
    //视野
    float V=4;
    //血量 预留
    int32 BloodValue=5;
}

//移动请求，其实就是个位置
message MovePackege
{
    Position P=1;
    int32 ActionData=2;
}

// 4 向外发送的广播
message BroadCast
{
    int32 Pid=1;
    //Type
    int32 Tp=2;
    //根据Tp内容不同，BroadCast消息会包含：
    //聊天内容(Content) 或初始位置(P) 或新位置(P*)
    //oneof类似于enum
    oneof Data
    {
        string Content=3;
        Position P=4;
        //ActionData暂时预留
        int32 ActionData=5;
    }
    string Username=6;
}

// 2 聊天信息
message Talk
{
    string Content=1;
}